src/core/fieldformatter/qgsrelationreferencefieldformatter.cpp

```diff
@@ -90,3 +93,62 @@ QVariant QgsRelationReferenceFieldFormatter::sortValue( QgsVectorLayer *layer, i
 {
   return representValue( layer, fieldIndex, config, cache, value );
 }
+
+QVariant QgsRelationReferenceFieldFormatter::createCache( QgsVectorLayer *layer, int fieldIndex, const QVariantMap &config ) const
+{
+  Q_UNUSED( fieldIndex );
+  QMap<QVariant, QString> cache;
+
+  // Some sanity checks
+  if ( !config.contains( QStringLiteral( "Relation" ) ) )
+  {
+    QgsMessageLog::logMessage( QObject::tr( "Missing Relation in configuration" ) );
+    return QVariant();
+  }
+  QgsRelation relation = QgsProject::instance()->relationManager()->relation( config[QStringLiteral( "Relation" )].toString() );
+  if ( !relation.isValid() )
+  {
+    QgsMessageLog::logMessage( QObject::tr( "Invalid relation" ) );
+    return QVariant();
+  }
+  QgsVectorLayer *referencingLayer = relation.referencingLayer();
+  if ( layer != referencingLayer )
+  {
+    QgsMessageLog::logMessage( QObject::tr( "representValue() with inconsistent layer parameter w.r.t relation referencingLayer" ) );
+    return QVariant();
+  }
+  QgsVectorLayer *referencedLayer = relation.referencedLayer();
+  if ( !referencedLayer )
+  {
+    QgsMessageLog::logMessage( QObject::tr( "Cannot find referenced layer" ) );
+    return QVariant();
+  }
+  int referencedFieldIdx = referencedLayer->fields().lookupField( relation.fieldPairs().at( 0 ).second );
+  if ( referencedFieldIdx == -1 )
+  {
+    QgsMessageLog::logMessage( QObject::tr( "Invalid referenced field (%1) configured in relation %2" ).arg( relation.fieldPairs().at( 0 ).second, relation.name() ) );
+    return QVariant();
+  }
+
+  QgsFeature feature;
+  auto iterator = referencedLayer->getFeatures();
```

Comment:
This request could be tightened to only get geometry when needed and only attributes required by the expression.